drake_ros_introspection/include/drake_ros_introspection/probes/port_probe.h

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <memory>
+#include <regex>
+#include <sstream>
+#include <string>
+
+#include <drake/common/drake_throw.h>
+#include <drake/systems/framework/context.h>
+#include <drake/systems/framework/event_status.h>
+#include <drake/systems/framework/framework_common.h>
+#include <drake_ros_introspection/probes/probe_interface.h>
+#include <drake_ros_introspection/utilities/port_traits.h>
+#include <drake_ros_introspection/utilities/type_conversion.h>
+#include <rclcpp/rclcpp.hpp>
+
+namespace drake_ros_introspection {
+namespace probes {
+
+template <typename PortT>
+class PortProbeInterface
+    : public ProbeInterface<typename utilities::port_traits<PortT>::ScalarT> {
+ public:
+  explicit PortProbeInterface(const PortT& port) : port_(port) {}
+
+  virtual ~PortProbeInterface() = default;
+
+  const PortT& get_port() { return port_; }
+
+ private:
+  const PortT& port_;
+};
+
+template <typename PortT, typename ValueT, auto ValueConvention,
+          typename MessageT>
+class PortProbe : public PortProbeInterface<PortT> {
+ public:
+  using PortProbeInterface<PortT>::PortProbeInterface;
+
+ private:
+  using ScalarT = typename utilities::port_traits<PortT>::ScalarT;
+
+  drake::systems::EventStatus DoProbe(
+      const drake::systems::Context<ScalarT>& context) override {
+    DRAKE_THROW_UNLESS(publisher_ != nullptr);
+    if (publisher_->get_subscription_count() > 0) {
+      const PortT& port = this->get_port();
+      const drake::systems::Context<ScalarT>& subcontext =
+          port.get_system().GetMyContextFromRoot(context);
+      using convert = utilities::convert<ValueT, ValueConvention, MessageT>;
+      publisher_->publish(
+          convert::to_message(port.template Eval<ValueT>(subcontext)));
+      return drake::systems::EventStatus::Succeeded();
```

Comment:
If the `Probe` isn't changing the state in the `Context`, mabye `DidNothing` is the appropriate return? The [docs say](https://drake.mit.edu/doxygen_cxx/classdrake_1_1systems_1_1_event_status.html#a11717da96dc4e4a5f79371318448f82b):

> kDidNothing | Successful, but nothing happened; no state update needed.

which makes me think returning `Succeeded` will cause `Simulator` to do extra work.

Reply:
Ah, good point. [It won't stop stepping](https://github.com/RobotLocomotion/drake/blob/a43f89bc58886452630eccfc15b4bbd5cc79d23f/systems/analysis/simulator.cc#L306) for as long as the monitor reports success (or time runs out).